### PR TITLE
GraalVM native-image build support

### PR DESCRIPTION
**Pull request by whilo**
This branch provides a prototype CLI tools built as a native image. The build uses https://github.com/taylorwood/clj.native-image. Preliminary documentation for the CLI can be found under `doc/cli.md`.

### CODE REVIEW

**Comment by jsmassa** [comment]
Needs an update to pass the tests.

**Comment by jsmassa** [comment]
Blocked by #542

**Comment by whilo** [comment]
@jsmassa Except for the backward-compatibility-test this should be ready to be merged now. Let me know what needs to be changed for this or feel free to fix it directly.

**Review by jsmassa — COMMENTED** [review]
Looks good to me. But as you say, it needs proper release versions for the dependencies and some benchmark checks

**Review by jsmassa — COMMENTED** [review]
Looks good to me in general. 

I am personally not a big fan of the package / directory name  'impl' though as it doesn't really say anything about the content. Since 'java' is taken, I guess I would choose something like 'interop' or 'native' in this case. But since we will have to do some restructuring anyway when changing to polylith, I am also ok with leaving it as is for now.

File: .circleci/config.yml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
             - deps-
       - run:
           name: resolve deps
-          command: clojure -P
+          command: clojure -A:dev:test:build -P
```
[the remark this answers is not in the document]
Comment (TimoKramer) [reply .circleci/config.yml:32]:
```
exec-opts:
 -Aaliases      Use concatenated aliases to modify classpath
 -X[aliases]    Use concatenated aliases to modify classpath or supply exec fn/args
 -T[toolname|aliases]  Invoke tool by name or via aliases ala -X
 -M[aliases]    Use concatenated aliases to modify classpath or supply main opts
 -P             Prepare deps - download libs, cache classpath, but don't exec
```
-A is fine if you only want to concatenate aliases. Here we need to cat them to prepare deps.

[the remark this answers is not in the document]
Comment (jsmassa) [reply .circleci/config.yml:32]: Ah, I see. Thanks for enlightening me :)

**Review by jsmassa — APPROVED** [review]

**Comment by whilo** [comment]
I agree, I strongly dislike `impl` as well unless it would disambiguate internals of something that is already scoped and then `internal` would probably be better. I actually tried `native` but `native` is a protected keyword in Java, so we would have to use something else. I inherited `impl` from libsci actually.

**Pull request merged**